### ISSUE TITLE
CS0246 build errors on .NET 6.0.300 SDK

### ISSUE THREAD

**Issue by mattjohnsonpint** [manuelroemer/Nullable#24]
After installing the .NET 6.0.300 SDK, doing `dotnet build --no-incremental` on a project using this library gives errors such as:

```
error CS0246: The type or namespace name 'NotNullWhen' could not be found
```

## Minimal Reproduction

MyApp.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <LangVersion>8</LangVersion>
    <Nullable>annotations</Nullable>
    <TargetFrameworks>net6.0;net48</TargetFrameworks>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Nullable" Version="1.3.0" PrivateAssets="all" />
  </ItemGroup>

</Project>
```

Program.cs
```csharp
using System;
using System.Diagnostics.CodeAnalysis;

namespace MyApp
{
    class Program
    {
        static void Main(string[] args)
        {
            if (Test(out var value))
            {
                Console.WriteLine(value);
            }
        }

        private static bool Test([NotNullWhen(true)] out string? value)
        {
            value = "Hello World!";
            return true;
        }
    }
}
```

Compile with `dotnet build --no-incremental`:

```
C:\Users\mattj\Code\MyApp>dotnet build --no-incremental
Microsoft (R) Build Engine version 17.2.0+41abc5629 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
C:\Users\mattj\Code\MyApp\Program.cs(16,35): error CS0246: The type or namespace name 'NotNullWhenAttribute' could not be found (are you m
issing a using directive or an assembly reference?) [C:\Users\mattj\Code\MyApp\MyApp.csproj]
C:\Users\mattj\Code\MyApp\Program.cs(16,35): error CS0246: The type or namespace name 'NotNullWhen' could not be found (are you missing a 
using directive or an assembly reference?) [C:\Users\mattj\Code\MyApp\MyApp.csproj]
  MyApp -> C:\Users\mattj\Code\MyApp\bin\Debug\net6.0\MyApp.dll

Build FAILED.
```

## Workaround
Install the previous 6.0.203 SDK, and force the project to use it by adding a `global.json` file with the following:
```json
{
  "sdk": {
    "version": "6.0.203"
  }
}
```

Then the project builds just fine with `dotnet build --no-incremental`.

This might be an upstream problem with .NET, but I haven't dug deep enough to figure out what it might be.

**Comment by mattjohnsonpint** [comment]
I also find it interesting that it only occurs with `--no-incremental`, even if this is the first build (no bin or obj folders).  Maybe that will help your investigation.

**Comment by shuebner** [comment]
I observed the failure in another smaller and simpler repo.
It built with 6.0.100, but not with 6.0.300.
The projects were NOT multitargeting, but targeting netstandard2.0.

**Comment by manuelroemer** [comment]
Hey @mattjohnsonpint, I just want to quickly thank you for opening the issue on the .NET SDK repo and all the others for helping to provide additional information on this issue. I appreciate you guys a lot!
Let's see what the outcome will be!

**Comment by manuelroemer** [comment]
I just merged #26 by @gtbuchanan and published the new package version 1.3.1 on NuGet. During my local tests the issue indeed seemed to be fixed (using .NET SDK 6.0.300). I will preemptively close this issue, but if someone still has any problems in this area even with the newer version, feel free to message me or simply reopen this issue.
Many thanks to all people involved (so far 😉)!
